keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+   - `private`: Default, disallow uncredentialed access to the storage instance.
+   - `ro`: Read only, uncredentialed users are permitted *read* operations for objects within the storage instance.
+   - `rw`: Read/Write, uncredentialed users are permitted *read and write* operations for objects within the storage instance. 
+   - `wo`: Write, uncredentialed users are permitted *write* operations for objects within the storage instance. 
+1. `bucketClassName`: Name of the associated BucketClass.
+1. `permittedNamespaces`: An array of namespaces, identified by a name and uid, from which  BucketRequests are allowed to bind to the Bucket.  Provided to allow admins a layer of cluster-layer access control.  Does **not** reflect or alter the backing storage instances' ACLs or IAM policies.
+   - In Greenfield, the originating BuckerRequest’s namespace must be specified at time of Bucket generation.
+   - In Brownfield, this list is defined by the admin.
+1. `protocol`: The protocol the application will use to access the storage instance.
+   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
+     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
+     - `s3`: data required to target a provisioned S3 bucket and/or user.
+     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
+1. `parameters`: a copy of the BucketClass parameters.
+1. `message`: a human readable description detailing the reason for the current `phase``.
+1. `phase`: is the current state of the Bucket:
+   - _Ready_: the controller finished processing and the controller is ready to receive access requests.
+   - _Released_: the originating Bucket has been deleted, signalling that the Bucket is ready for garbage collection.  This will occur on Greenfield Buckets once all requests referencing the Bucket are deleted.
+   - _Failed_: error and all retries have been exhausted.
+   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.
+
+#### BucketClass
+
+A cluster-scoped API whose purpose it to provide admins control over the handling of bucket provisioning by specifying provisioner-defined key-values in the `parameters` map.  The `BucketClass` defines a release policy, and specifies driver specific parameters and the provisioner name.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: 
+provisioner: [1]
+isDefaultBucketClass: [2]
+supportedProtocols: {"azureblob", "gcs", "s3", ... } [3]
+anonymousAccessMode: {"ro", "wo", "rw"} [4]
+additionalPermittedNamespaces: [5]
+- name:
+  uid: 
+releasePolicy: {"Delete", "Retain"} [6]
+parameters: [7]
+```
+
+1. `provisioner`: The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.
```

Comment:
What _secretRef_ does this refer to?
And, is static provisioning supported?  
@wlan0

Reply:
UUIC, staticProvisioning is where buckets are created manually with specified names rather than where Kubernetes auto-generates a name and creates it on the fly.